mava/wrappers/gigastep.py

```diff
@@ -0,0 +1,300 @@
+            state (GigastepState): The current state of the environment.
+            action (Array): The actions for controllable agents.
+
+        Returns:
+            Tuple[GigastepState, TimeStep]: A tuple containing the next state of the environment
+            and the next time step.
+
+        """
+        key, step_key, adversary_key = jax.random.split(state.key, 3)
+
+        action = jnp.concatenate([action, state.adversary_action], axis=0, dtype=jnp.int16)
+
+        obs, env_state, rewards, dones, ep_done = self._env.step(state.state, action, step_key)
+
+        # cut out the rewards,dones of the adversary
+        rewards, dones = (
+            rewards[: self.num_agents],
+            dones[: self.num_agents],
+        )
+
+        obs_team1, state_team1, obs_team2, state_team2 = self._split_obs_and_state(obs, env_state)
+
+        # take the actions of the adversary and cache it before returning the new state
+        adversary_actions = self.adversary_policy(obs_team2, state_team2, adversary_key)
+
+        obs = self._create_observation(obs_team1, adversary_actions, obs, state)
+
+        step_type = jax.lax.select(ep_done, StepType.LAST, StepType.MID)
+
+        current_winner = ep_done & self.won_episode(env_state)
+
+        ts = TimeStep(
+            step_type=step_type,
+            reward=rewards,
+            discount=1.0 - dones,
+            observation=obs,
+            extras={"won_episode": current_winner},
+        )
+        return GigastepState(env_state, key, state.step + 1, adversary_actions), ts
+
+    def _create_observation(
+        self,
+        obs: Array,
+        adversary_actions: Array,
+        obs_full: Array,
+        state: GigastepState,
+    ) -> Union[Observation, ObservationGlobalState]:
+        """Create an observation from the raw observation and environment state."""
+        obs_data = {
+            "agents_view": obs,
+            "action_mask": self.action_mask(),
+            "step_count": jnp.repeat(state.step, self.num_agents),
+        }
+
+        if self.has_global_state:
+            obs_data["global_state"] = self.get_global_state(obs_full, adversary_actions)
+            return ObservationGlobalState(**obs_data)
+        else:
+            return Observation(**obs_data)
+
+    def action_mask(self) -> Array:
+        """Get action mask for each agent."""
+        return jnp.ones((self.num_agents, self._env.n_actions))  # all actions are valid
+
+    def get_global_state(self, obs: Array, adversary_actions: Array) -> Array:
+        """
+        Combines observations from all agents,
+        adversary actions, and adversary observations
+        to create a global state for the environment.
+
+        Args:
+            obs (Array): The observations of all agents.
+            adversary_actions (Array): The actions for the adversary.
+
+        Returns:
+            global_obs (Array): The global observation.
+        """
+        global_obs = jnp.concatenate(obs, axis=0)
```

Comment:
As discussed lets add a comment here to remind us to check this

```suggestion
        # todo: still unsure if adversary actions and observations is too much information in 
        # the global Needs to be tested once we have better heuristics for adversaries.
        global_obs = jnp.concatenate(obs, axis=0)
```